src/crypto/bip324_suite.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2019-2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/bip324_suite.h>
+
+#include <crypto/common.h>
+#include <crypto/poly1305.h>
+#include <crypto/sha256.h>
+#include <hash.h>
+#include <span.h>
+#include <support/cleanse.h>
+
+#include <assert.h>
+#include <cstring>
+#include <string.h>
+
+#ifndef HAVE_TIMINGSAFE_BCMP
+
+int timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
+{
+    const unsigned char *p1 = b1, *p2 = b2;
+    int ret = 0;
+
+    for (; n > 0; n--)
+        ret |= *p1++ ^ *p2++;
+    return (ret != 0);
+}
+
+#endif // TIMINGSAFE_BCMP
+
+BIP324CipherSuite::~BIP324CipherSuite()
+{
+    memory_cleanse(payload_key.data(), payload_key.size());
+    memory_cleanse(rekey_salt.data(), rekey_salt.size());
+}
+
+void BIP324CipherSuite::CommitToKeys(const Span<const std::byte> data, bool commit_to_L, bool commit_to_P)
+{
+    if (commit_to_L) {
+        fsc20.CommitToKey(data);
+    }
+
+    if (commit_to_P) {
+        HashWriter hasher;
+        assert(CSHA256::OUTPUT_SIZE == BIP324_KEY_LEN);
+        hasher << MakeUCharSpan(rekey_salt) << MakeUCharSpan(data) << MakeUCharSpan(payload_key);
+        auto new_key = hasher.GetSHA256();
+        memcpy(payload_key.data(), new_key.data(), BIP324_KEY_LEN);
+    }
+
+    rekey_ctr++;
+    msg_ctr = 0;
+    set_nonce();
+}
+
+bool BIP324CipherSuite::Crypt(const Span<const std::byte> input, Span<std::byte> output,
+                              BIP324HeaderFlags& flags, bool encrypt)
+{
+    // check buffer boundaries
+    if (
+        // if we encrypt, make sure the destination has the space for the length field, header, ciphertext and MAC
+        (encrypt && (output.size() < BIP324_LENGTH_FIELD_LEN + BIP324_HEADER_LEN + input.size() + RFC8439_TAGLEN)) ||
+        // if we decrypt, make sure the source contains at least the header + mac and the destination has the space for the source - MAC - header
+        (!encrypt && (input.size() < BIP324_HEADER_LEN + RFC8439_TAGLEN || output.size() < input.size() - BIP324_HEADER_LEN - RFC8439_TAGLEN))) {
+        return false;
+    }
+
+    if (encrypt) {
+        // input is just the payload
+        // output will be encrypted length + encrypted (header and payload) + mac tag
+        uint32_t ciphertext_len = BIP324_HEADER_LEN + input.size();
+        WriteLE32(reinterpret_cast<unsigned char*>(&ciphertext_len), ciphertext_len);
+
+        std::vector<std::byte> input_vec;
+        input_vec.resize(BIP324_HEADER_LEN + input.size());
```

Comment:
361e971: `input_vec` can be defined directly with the required size.

Reply:
Fixed.